fish-rust/src/builtins/ulimit.rs

```diff
@@ -0,0 +1,354 @@
+use std::cmp::max;
+
+use libc::{c_int, RLIM_INFINITY};
+use nix::{
+    errno::Errno::EPERM,
+    sys::resource::{self, getrlimit, setrlimit},
+};
+
+use crate::{
+    builtins::shared::STATUS_CMD_ERROR,
+    ffi::{fish_wcswidth2, parser_t},
+    wchar::{wstr, L},
+    wchar_ffi::WCharToFFI,
+    wgetopt::{wgetopter_t, wopt, woption, woption_argument_t::no_argument},
+    wutil::wgettext_fmt,
+};
+
+use super::shared::{builtin_wperror, io_streams_t, STATUS_CMD_OK};
+
+/// Struct describing a resource limit.
+#[derive(Debug, Copy, Clone)]
+struct Resource<'a> {
+    /// resource ID
+    resource: resource::Resource,
+    /// description of resource
+    desc: &'a wstr,
+    /// switch used on commandline to specify resource
+    switch_char: char,
+    /// the implicit multiplier used when setting getting values
+    multiplier: i32,
+}
+
+/// Array of `Resource` structs, describing all known resource types.
+const resource_arr: &[Resource] = &[
```

Comment:
I'm not sure if you intend for this to be `const` and not `static`. `ulimit` isn't a hotspot in the code so I would prefer to prevent allocating space/memory beforehand (i.e. `const`) but at the same time making it `static` might end up  with better overall codegen.

Also, please use idiomatic naming and use screaming snake case here.

Reply:
> ulimit isn't a hotspot in the code so I would prefer to prevent allocating space/memory beforehand

The data needs to be part of the binary anyway, adding a `static` (`LazyCell` or similar I guess?) would be strictly more expensive because the data needs to be copied to a heap allocation, no?

Reply:
Seconded screaming snake case (`RESOURCE_ARR`). The use of the static const here is fine, it matches the C++ and there's no unnecessary allocations.